### ISSUE TITLE
[laemmel] Coupled Software Transformations

### ISSUE THREAD

**Issue by grammarware** [grammarware/sle16#14]
Submitted by @rlaemmel to https://github.com/grammarware/sle16/tree/master/laemmel

**Comment by rlaemmel** [comment]
Ok, I saw that Vadim was unarchiving the zip file so that now all files are now "native" files in this repo. 

This triggers the question, which perhaps the reviewer could kindly try to answer, as to whether it may be better to bind all the paper URLs to this repo rather than to the primary YAS repo. 

Pros:
- No changes are applied to the sle16 repo; so things are more reproducible?

Cons:
- If the primary artefacts (in the primary repo location) get improved, then this will be less discoverable.

**Comment by grammarware** [comment]
@rlaemmel, there are several concerns here. One is making an honest snapshot of the work in the stage it was during the publication — we hope to address that by not just reviewing the artefacts, but also asking authors of **approved** ones to add them to their publication as auxiliary material. In this case, ACM as a third party will take care of storing the snapshots. In this respect, this repository is a temporary platform to facilitate reviewing.

Another concern is that most artefacts at SLE are executables (which is not the case for all CS/SE conferences, but here it seems to be), and executable software tend to evolve and get changed due to maintenance activities. This could potentially be addressed by linking the SPLASH/SLE website to this repo's READMEs and keeping the repo up indefinitely with authors having access to it. Alternatively, we can just link **directly** to live repos.

We are already discussing another option with authors of a paper that received a lot of constructive feedback — in fact, so much that they are changing the implementation significantly. This is a good thing, but it essentially circles them out of the artefact evaluation process since they do not have a good snapshot at this moment, but will have it during the conference or shortly before it. We were contemplating doing another round of artefact... hmm... perhaps not evaluation, but collection, a week or two before the conference, to enhance the list of papers at the SPLASH/SLE website with links to downloadable artefacts.

**Comment by grammarware** [comment]
As for the original question of where should the PDF link to: we'll see what the reviewers say...

**Comment by rlaemmel** [comment]
Well, as Vadim's says that the sle16 repo is more of a tool for reviewing, then I definitely prefer keeping links to the live repo, as it should remain up and running for some time. Of course, I might end up wanting to break the file naming pattern or just individual file names; that's what worries me. I could also just create a designated repo for "my sle paper" and freeze a version in it, or perhaps just a branch -- I assume I can link into branches.

**Comment by grammarware** [comment]
Yep, you can refer to branches/commits and files within branches/commits.

e.g., https://github.com/grammarware/slps/blob/zoo/shared/bgf/ada-laemmel-verhoef.bgf or https://github.com/grammarware/slps/blob/5b2825e4b8fd327b8cf928e99ecc469eeb33f749/shared/bgf/Makefile

```
https://github.com/grammarware/slps/blob/zoo/shared/bgf/ada-laemmel-verhoef.bgf
                                                       ^^^^^^^^^^^^^^^^^^^^^^^^ filename
                                            ^^^^^^^^^^^^ path
                                        ^^^^^ branch name
                              ^^^^^^ repository
                  ^^^^^^^^^^^^^ owner

https://github.com/grammarware/slps/blob/5b2825e4b8fd327b8cf928e99ecc469eeb33f749/shared/bgf/Makefile
                                                                                 ^^^^^^^^^^^^ path
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ commit hash
                              ^^^^^^ repository
                  ^^^^^^^^^^^^^ owner
```

**Comment by patrimrodilla** [comment]
Dear all,
I followed the readme file instructions and the artifact is failing in the execution. After the lines indicated by the authors as indication of successfully make (103 languages, 552 files, etc.). The execution shows an error. The error is in the attached screenshot, said "Megamodel problems: Baseline languages/ueber/functions.term: disagreeing" and the execution aborted. Please, take a look in order to guess what happens and if it's only in my case or is a general crash,

Thank you!

![captura de pantalla 2016-09-08 a las 11 37 56](https://cloud.githubusercontent.com/assets/21332357/18344929/98dd7552-75b9-11e6-9a4d-baf08db420bf.png)

**Comment by iovinoludovico** [comment]
Dear all, I've executed the process as indicated and in my case after installing the pre-requisites on my laptop (prolog), the workflow is terminated without errors. I attach the screen ... for @patrimrodilla  maybe the errors comes from the prolog version?
![schermata 2016-09-08 alle 19 01 04](https://cloud.githubusercontent.com/assets/1259888/18358790/1e465668-75f7-11e6-88a8-5e51488e600f.png)

**Comment by iovinoludovico** [comment]
I have been looking for an intro on the website but at the end I started to understand inspecting the makefile and the paper. So my first consideration is to report some of the recommendations for the first explanation and execution also in the repo or the website. Of course i understand that the tool requires some background and only reading the paper can explain how the megamodeling is represented with this approach. In my opinion the provided artefacts are consistent with the paper since the paper reports the exact path of the described file; but in my case (also considering my poor expertise in these tools) the tool is not documented well or maybe I've not been able to find the proper link. The tool is not easy to reuse for me, but again I highlights that could be related to the reviewer expertise. The approach presented in the paper also include bidirectionality that increases the complexity of the examples.

**Comment by patrimrodilla** [comment]
Dear all,

thank you @iovinoludovico for the feedback. It's working now with the Prolog v 7.2.3. However, I recommend if any of you have The Captain as MacOS version and you have problems (I'll check in other laptop with The Captain and it's ok, so maybe it's a software configuration issue), installing the Prolog development version 7.3 or later, instead the recommended version by the authors. I attached the screenshot of WARNING Prolog message to clarify.
![captura de pantalla 2016-09-07 a las 14 46 37](https://cloud.githubusercontent.com/assets/21332357/18359371/ee533748-75f9-11e6-8f96-c0175f9a1835.png)

Anyway, this kind of Prolog dependencies are not good for any artifact, tricky third-part dependencies...

Thanks again,

**Comment by patrimrodilla** [comment]
After the small problems referred to Prolog dependencies, artifacts (becasue constitutes a collection of files) were processed correctly. I agree with @iovinoludovico that perhaps the initial **documentation is a bit limited** in the repository (not in the paper), as you may expect a prototype as an artifact, but actually they are the different files that illustrate the contributions of the paper.

Saving this problem in the documentation (which would be easy fix explaining it in the readme file or similar solution), artifacts are **consistent** with the paper presented, keeping the paths/links to them.
They are also **complete** in terms of developing the LAL language, the CX patterns presented and the semantics translation.

I have, however, some concerns in terms of the reusability of artifacts. Of course, I understand that the contributions of the paper itself represent a way to use the concept of megamodel in order to represent and work with this type of transformations, which is reusable in several problems and domains. However, I have not so clear is these artifacts themselves are reusable, mainly due to the existence of entry barriers/ or a high learning curve (to know the LAL language, for example, among others). This is a direct consequence also in not developing a prototype as demostratos artifact itself, but only present as artifact the original products/contributions of the paper. For future work, I could suggest to the authors trying to eliminate this problem by implementing this kind of artifact and use it in dissemination terms.

Technological dependencies, as I said in previous posts, does also not help very much to the re-use capabilities of the artifact by other professionals.

In short, the artifacts needs improvements in documentation for dissemination and re-use capabilities, although they are complete and consistent.

**Comment by altmattr** [comment]
I'm attempting to review this artefact on linux and there is a small problem due to the case-sensitive file system there.  The following patch fixes it for me and gets me further along (to an issue similar to @patrimrodilla)  I am offering the patch in-case anyone else is on a case-sensitive file system

```
--- a/laemmel/yas/languages/BGL/api.pro
+++ b/laemmel/yas/languages/BGL/api.pro
@@ -68,7 +68,7 @@ parserOfBgl(Text, Term2) :-
   require(
     scannerFailed(egl),
     tokens(eglToken, Text, Tokens)),
-  File = 'languages/egl/cs.term',
+  File = 'languages/EGL/cs.term',
   require(
     fileNotReadable(File),
     readTermFile(File, Egl)),
```

```
--- a/laemmel/yas/languages/BSL/api.pro
+++ b/laemmel/yas/languages/BSL/api.pro
@@ -14,11 +14,11 @@ bslLanguage(File, Term) :-

 % Parser of BSL signatures relying on ESL
 parserOfBsl(Text, Term2) :-
-  FileCfSyn = 'languages/esl/cs.term',
+  FileCfSyn = 'languages/ESL/cs.term',
   require(
     fileNotReadable(FileCfSyn),
     readTermFile(FileCfSyn, CfSyn)),
-  FileLexSyn = 'languages/esl/ls.term',
+  FileLexSyn = 'languages/ESL/ls.term',
   require(
     fileNotReadable(FileLexSyn),
     readTermFile(FileLexSyn, LexSyn)),
```

**Comment by altmattr** [comment]
Iv'e successfully tested the artefact and it is complete and consistent.  Thus I am happy to recommend the artefact for acceptance.

As well as the above case-sensitivity blocker, I was only able to run the artefact on MacOSX with the precisely defined version of Prolog.  The same version of Prolog from the debian repository and a different version on MacOS both caused errors.  This fragility to versions means the artefact is not as re-usable as one would like.  Were the artefact dependent on a freely available (linux for example) OS, the problem would be lessened but there are now, and will certainly be in the future, some people who can't run this for lack of access to hardware.  It seems to be subtle prolog semantic inconsistencies and I expect the fix is possible with a little more effort than I was able to provide here.

All files/definitions needed to replicate and verify the results in the paper are present and the artefact is well organised/coded.  There is an incredible amount of detail in the code but I was able to trace lines of inquiry through the code to see how various parts worked.

**Comment by rlaemmel** [comment]
cc @grammarware - I plan to release the artefacts via a branch of the underlying github repo, as we discussed on GitHub previously. Does this committee have any additional constraints on how to do this? As far as I understand, https://github.com/grammarware/sle16 is not the final repo for the artefacts anyway, right? How is the SLE conference ultitematly planning to refer to the "final" artefacts?

cc @altmattr - thanks for your comments on completeness and consistency. I will definitely pick up your input in finalizing the artefacts so that we get better compatibility on Linux and across Prolog versions.

cc @patrimrodilla - you say "In short, the artifacts needs improvements in documentation for dissemination and re-use capabilities, although they are complete and consistent." I think that the paper at hand is designed to disseminate the artefacts at hand. It clearly lists inputs, outputs, and parameters. What I could do is that I provide a simple example with entries like this, all with links to sample files:
- Megamodel
- Sample transformation
- Samples transformation input
- Samples transformation output
- Test-case description
- Console capture of executing test cases
  I hope to add this in the next two days.
  Still thinking of how to better address your concern.

**Comment by iovinoludovico** [comment]
I agree with Ralf's proposal of providing a simple example with the suggested artifacts. This will improve the understandability and usability of the tool.

**Comment by grammarware** [comment]
The Artefact Evaluation Committee of SLE 2016 has reached the conclusion yesterday during the committee meeting that this artefact conforms to the expectations and is **approved** for inclusion as auxiliary material for the publication. However, we would like to stress that the complexity level of this artefact was definitely above the median complexity the committee members have observed in other artefacts, so any guidance you can provide to walk the future potential reusers/replicators of your artefact, is appreciated.

The final artefacts are submitted as auxiliary materials attached to the papers submitted to the ACM. If you think this is inappropriate for your artefact, you can archive them yourself (say, in a branch) and link to if from the paper — in this case you are still allowed and encouraged to add the badge of approval. The upside of this approach is the same as its downside — you will be responsible for maintaining the final artefact resting place forever.

Please make some appropriate adjustments before submitting the materials to ACM, @rlaemmel, and enjoy your seal of approval.

> PS: I, @grammarware, have a conflict of interest with the author of this submission, @rlaemmel, so I have intentionally not participated to any extent in the assessment and discussion of this artefact, except for posting this notification, which is a purely administrative act of delivering the decision that has already been taken.

**Comment by rlaemmel** [comment]
Thanks everyone for help.

Here is what I did:
- The code has been singled out and frozen in a branch: https://github.com/softlang/yas/tree/sle16
- The revised, pre-final paper links to the new branch: http://softlang.uni-koblenz.de/cxrevisited/
- The revised README (for now only at new branch, not here) links to files of the key CX example.
- I fixed the case-related compatibility problems reported.

Hoping to get more Linux platform testing completed by the deadline, but thanks for not making this a requirement this time around.
